File: games/driving/tower.ts
// =============================================================================
// tower — the radio tower that stands beyond each intersection: a tall square-base
// lattice pyramid of metal rods. The first of the "truly 3D" landmarks an
// intersection owns (an intersection is a PLACE, not just a junction). This module
// owns the tower's dimensions and how it draws; the intersection only places it, out past the
// corner and off to the right, in the direction the Rider faces as he approaches (the same
// "beyond the turn, in his facing direction" idea the corner creatures' baby uses).
//
// Cheap by construction: a see-through wire lattice of flat rod-quads (no solid faces).
// It picks its OWN level of detail, since a tower is never within the renderer's critter-tuned
// DETAIL_DIST: a full perspective lattice within TOWER_NEAR_DIST, and beyond that the SAME four
// faces drawn as a flat billboard (project the base corners + apex, interpolate the rest in 2D).
// A far tower has no perceptible depth, so the flat version reads identically — and because it
// keeps all four faces, the back leg doesn't pop into view as a tower crosses the threshold. The
// flat path also skips the per-vertex frame-mapping and near-clip the 3D one needs.
// =============================================================================

import { NEAR } from './scenery.ts';
import type { Project, Ctx, Scenery, RiderPt } from './scenery.ts';

// ---- dimensions (metres) ----
const TOWER_HEIGHT = 80;               // apex height
const TOWER_HALF = 6;                  // half the 12m square base edge
const TOWER_BEYOND = 160;             // how far past the corner it stands, along the approach direction
const TOWER_RIGHT = 20;              // offset to the right of the lane, so you don't bear down on it dead-on
const STAGE_HEIGHT = 20;              // a cross-beam ring every this many metres (rings at 20/40/60)
const BRACE_STAGES = 2;              // X-braces only on the bottom this-many stages (cheaper, and where they matter structurally)
const ROD_HALF = 0.12;               // half the rod thickness
const TOWER_YAW = 30 * Math.PI / 180; // turn the square off head-on so the Rider sees two faces, not one

// Beyond this range a tower switches from the perspective lattice to the flat billboard. Both
// draw all four faces, so nothing pops — the only thing lost is parallax, imperceptible this far off.
const TOWER_NEAR_DIST = 400;

const TOWER_METAL = '#9aa0a8';   // every rod — legs, rings, and braces — one darker gray

// ---- the beacon at the apex ----
const BEACON_PERIOD = 120;       // frames for a full invisible -> bright -> invisible cycle
const BEACON_RADIUS = 0.5;       // 1m-diameter sphere, drawn as a flat disc
const BEACON_COLOR = '#ff7a18';  // bright orange at full brightness

// The blink is a pure function of the step (used as a clock), so it's identical on pause and
// runs backwards on reverse — no wall-clock state. Each tower gets an authored phase offset so
// they don't blink in unison. brightness: 0 (invisible) at the cycle ends, 1 (full) at the middle.
export function beaconOffsetFor(segNum: number): number {
  return (segNum * 37) % BEACON_PERIOD;   // 37 is coprime to 120 — spreads the offsets, all distinct
}
function beaconBrightness(step: number): number {
  const phase = ((step % BEACON_PERIOD) + BEACON_PERIOD) % BEACON_PERIOD;
  return (1 - Math.cos(2 * Math.PI * phase / BEACON_PERIOD)) / 2;
}

// a rider-frame point carrying a height off the ground
interface Pt3 { right: number; forward: number; height: number }
interface Rod { pts: Pt3[]; color: string }
interface Pt2 { x: number; y: number }   // a projected screen point (the flat path works in 2D)

// Clip a rod polygon against the near plane (forward >= NEAR) in 3D, before projecting —
// the same Sutherland-Hodgman the renderer runs on road quads, here on height-carrying points.
function clipNear(pts: Pt3[]): Pt3[] {
  const out: Pt3[] = [];
  for (let i = 0; i < pts.length; i++) {
    const a = pts[i], b = pts[(i + 1) % pts.length];
    const aIn = a.forward >= NEAR, bIn = b.forward >= NEAR;
    if (aIn) out.push(a);
    if (aIn !== bIn) {
      const f = (NEAR - a.forward) / (b.forward - a.forward);
      out.push({
        right: a.right + f * (b.right - a.right),
        forward: NEAR,
        height: a.height + f * (b.height - a.height),
      });
    }
  }
  return out;
}

// the square's four base corners, in half-base units, before the yaw
const CORNERS = [[-1, -1], [1, -1], [1, 1], [-1, 1]];

// Build the tower standing beyond the intersection whose incoming segment is `fromLength`
// long. `map(a, x)` takes a point in that segment's BL frame (a along, x across-from-left) to
// the Rider's frame — the same mapper the intersection uses for its other scenery. `step` is the
// frame clock and `beaconOffset` this tower's authored blink phase (see beaconBrightness).
export function towerScenery(map: (a: number, x: number) => RiderPt, fromLength: number, hw: number, step: number, beaconOffset: number): Scenery {
  const a0 = fromLength + TOWER_BEYOND;   // out ahead of the corner, in the approach direction
  const x0 = hw + TOWER_RIGHT;            // and off to the right (BL across = half-width is the centreline)
  const cy = Math.cos(TOWER_YAW), sy = Math.sin(TOWER_YAW);

  // corner k of the cross-section at height h, in the Rider's frame. The square shrinks
  // linearly to a point at the apex, and is yawed about the vertical axis.
  const at = (k: number, h: number): Pt3 => {
    const s = TOWER_HALF * (1 - h / TOWER_HEIGHT);
    const du = CORNERS[k][0] * s, dv = CORNERS[k][1] * s;
    const ru = du * cy - dv * sy, rv = du * sy + dv * cy;   // yaw
    const p = map(a0 + rv, x0 + ru);
    return { right: p.right, forward: p.forward, height: h };
  };

  // a rod as a flat thin ribbon between two points. Sloping rods (legs, diagonal braces) take
  // their width horizontally (offset in `right`); horizontal beams take it vertically (offset in
  // height) — whichever direction reads as thickness from the road.
  const barH = (a: Pt3, b: Pt3, color: string): Rod => ({ color, pts: [
    { right: a.right - ROD_HALF, forward: a.forward, height: a.height },
    { right: b.right - ROD_HALF, forward: b.forward, height: b.height },
    { right: b.right + ROD_HALF, forward: b.forward, height: b.height },
    { right: a.right + ROD_HALF, forward: a.forward, height: a.height },
  ] });
  const barV = (a: Pt3, b: Pt3, color: string): Rod => ({ color, pts: [
    { right: a.right, forward: a.forward, height: a.height - ROD_HALF },
    { right: b.right, forward: b.forward, height: b.height - ROD_HALF },
    { right: b.right, forward: b.forward, height: b.height + ROD_HALF },
    { right: a.right, forward: a.forward, height: a.height + ROD_HALF },
  ] });

  const apex = at(0, TOWER_HEIGHT);   // s = 0, so every corner converges here
  const center = map(a0, x0);

  // ---- NEAR: the full perspective lattice (built lazily — a far tower never pays for it) ----
  // Four sloping legs, horizontal cross-beam rings (the top stage tapers to the apex, so no ring
  // there), and X-braces on the bottom BRACE_STAGES stage faces, both diagonals crossing into an X.
  const lattice = (): Rod[] => {
    const rods: Rod[] = [];
    for (let k = 0; k < 4; k++) rods.push(barH(at(k, 0), apex, TOWER_METAL));
    for (let h = STAGE_HEIGHT; h < TOWER_HEIGHT; h += STAGE_HEIGHT) {
      for (let k = 0; k < 4; k++) rods.push(barV(at(k, h), at((k + 1) % 4, h), TOWER_METAL));
    }
    for (let h = 0; h < BRACE_STAGES * STAGE_HEIGHT; h += STAGE_HEIGHT) {
      const hi = h + STAGE_HEIGHT;
      for (let k = 0; k < 4; k++) {
        const j = (k + 1) % 4;
        rods.push(barH(at(k, h), at(j, hi), TOWER_METAL));
        rods.push(barH(at(j, h), at(k, hi), TOWER_METAL));
      }
    }
    return rods;
  };
  // back-to-front so the near face's rods overpaint the far face's.
  const avgF = (r: Rod): number => (r.pts[0].forward + r.pts[1].forward + r.pts[2].forward + r.pts[3].forward) / 4;
  const sortBack = (rods: Rod[]): Rod[] => [...rods].sort((p, q) => avgF(q) - avgF(p));

  const drawNear = (ctx: Ctx, project: Project): void => {
    for (const rod of sortBack(lattice())) {
      // near-clip first: a rod that straddles the near plane (one corner beside/behind the
      // Rider on a sharp turn) would otherwise project to a sliver streaking across the view.
      const pts = clipNear(rod.pts);
      if (pts.length < 3) continue;
      ctx.fillStyle = rod.color;
      ctx.beginPath();
      const s0 = project(pts[0].right, pts[0].forward, pts[0].height);
      ctx.moveTo(s0.x, s0.y);
      for (let i = 1; i < pts.length; i++) {
        const s = project(pts[i].right, pts[i].forward, pts[i].height);
        ctx.lineTo(s.x, s.y);
      }
      ctx.closePath();
      ctx.fill();
    }
  };

  // ---- FAR: the same four faces, flat. Project the base corners + apex once, then derive every
  // ring corner by 2D interpolation up the legs (the cross-section shrinks linearly to the apex).
  // No parallax, but all four faces, so the back leg is present at the threshold just like near. ----
  const ROD_W = ROD_HALF * 2;
  const drawFlat = (ctx: Ctx, project: Project): void => {
    const ps = (p: Pt3): Pt2 => project(p.right, p.forward, p.height);
    const baseS = [ps(at(0, 0)), ps(at(1, 0)), ps(at(2, 0)), ps(at(3, 0))];
    const apexS = ps(apex);
    const wpx = ROD_W * (project(1, center.forward, 0).x - project(0, center.forward, 0).x);   // rod width in px at this depth
    const lerp = (a: Pt2, b: Pt2, t: number): Pt2 => ({ x: a.x + (b.x - a.x) * t, y: a.y + (b.y - a.y) * t });
    const ring = (k: number, h: number): Pt2 => lerp(baseS[k], apexS, h / TOWER_HEIGHT);
    const bar = (A: Pt2, B: Pt2): void => {   // a thick screen-space line as a filled quad
      const dx = B.x - A.x, dy = B.y - A.y, len = Math.hypot(dx, dy) || 1;
      const ox = -dy / len * wpx / 2, oy = dx / len * wpx / 2;
      ctx.beginPath();
      ctx.moveTo(A.x + ox, A.y + oy); ctx.lineTo(B.x + ox, B.y + oy);
      ctx.lineTo(B.x - ox, B.y - oy); ctx.lineTo(A.x - ox, A.y - oy);
      ctx.closePath();
      ctx.fill();
    };
    ctx.fillStyle = TOWER_METAL;
    for (let k = 0; k < 4; k++) bar(baseS[k], apexS);                                  // legs
    for (let h = STAGE_HEIGHT; h < TOWER_HEIGHT; h += STAGE_HEIGHT) {
      for (let k = 0; k < 4; k++) bar(ring(k, h), ring((k + 1) % 4, h));               // rings
    }
    for (let h = 0; h < BRACE_STAGES * STAGE_HEIGHT; h += STAGE_HEIGHT) {
      const hi = h + STAGE_HEIGHT;
      for (let k = 0; k < 4; k++) { const j = (k + 1) % 4; bar(ring(k, h), ring(j, hi)); bar(ring(j, h), ring(k, hi)); }
    }
  };

  // ---- the apex beacon: a flat orange disc whose SIZE scales with distance but whose
  // BRIGHTNESS (alpha) does not — a depthless glow that blinks on the step clock. ----
  const beacon = beaconBrightness(step + beaconOffset);
  const drawBeacon = (ctx: Ctx, project: Project): void => {
    if (beacon < 0.02) return;   // skip the invisible part of the cycle
    if (apex.forward < NEAR) return;
    const s = project(apex.right, apex.forward, apex.height);
    const r = BEACON_RADIUS * (project(1, apex.forward, 0).x - project(0, apex.forward, 0).x);
    ctx.globalAlpha = beacon;
    ctx.fillStyle = BEACON_COLOR;
    ctx.beginPath();
    ctx.arc(s.x, s.y, r, 0, 2 * Math.PI);
    ctx.fill();
    ctx.globalAlpha = 1;
  };

  const draw = (ctx: Ctx, project: Project): void => {
    (center.forward < TOWER_NEAR_DIST ? drawNear : drawFlat)(ctx, project);
    drawBeacon(ctx, project);
  };

  return { forward: center.forward, height: TOWER_HEIGHT, drawAsNear: draw, drawAsFar: draw };
}

File: games/driving/wasm/tower.zig
//! tower — the radio tower beyond each intersection (and mid-way down a long
//! segment): a tall square-base lattice pyramid of metal rods, tapering to an apex
//! with a pink beacon. The owner (render, via the world) places it; this module only
//! renders. Mirrors tower.ts.
//!
//! For now this is the FLAT form only — the four faces projected as a 2D billboard
//! (base corners + apex projected, every ring/brace derived by screen-space lerp),
//! which is exactly what tower.ts draws beyond TOWER_NEAR_DIST. The near 3D-parallax
//! lattice and the step-driven beacon blink are deferred; the beacon is drawn static.
//! Local ground curvature (EARTH_RADIUS) sinks a far tower so its base meets the
//! horizon and its apex pulls down.

const std = @import("std");
const camera = @import("camera.zig");
const geom = @import("geom.zig");
const paint = @import("paint.zig");

const TOWER_HEIGHT: f32 = 80; // apex height (m)
const TOWER_HALF: f32 = 6; // half the 12m square base edge
const STAGE_HEIGHT: f32 = 20; // a cross-beam ring every this many metres (20/40/60)
const BRACE_STAGES: usize = 2; // X-braces only on the bottom this-many stages
const ROD_HALF: f32 = 0.12;
const ROD_W: f32 = ROD_HALF * 2;
const TOWER_METAL: u32 = 0x9aa0a8;
const EARTH_RADIUS: f32 = 20000; // local ground bulge (towers only — stronger than the road's)
const BEACON_RADIUS: f32 = 3.0;
const BEACON_COLOR: u32 = 0xff2fe6;

// the square's four base corners, in half-base units, before the yaw.
const CORNERS = [4][2]f32{ .{ -1, -1 }, .{ 1, -1 }, .{ 1, 1 }, .{ -1, 1 } };

/// baseCornerAX returns base corner k's position in the owning segment's BL frame
/// (`a` along, `x` across-from-left), given the tower's base centre + yaw. The caller
/// maps it into the rider frame. (At the apex the square shrinks to a point at the
/// centre, so the apex is just map(a0, x0) at TOWER_HEIGHT — no helper needed.)
pub fn baseCornerAX(k: usize, a0: f32, x0: f32, yaw: f32) geom.AX {
    const du = CORNERS[k][0] * TOWER_HALF;
    const dv = CORNERS[k][1] * TOWER_HALF;
    const cy = @cos(yaw);
    const sy = @sin(yaw);
    const ru = du * cy - dv * sy;
    const rv = du * sy + dv * cy;
    return .{ .a = a0 + rv, .x = x0 + ru };
}

fn groundDrop(p: geom.RiderPt) f32 {
    return (p.right * p.right + p.forward * p.forward) / (2.0 * EARTH_RADIUS);
}

fn proj(right: f32, forward: f32, h: f32, drop: f32, cam_focal: f32) camera.ScreenPt {
    return camera.project(.{ .right = right, .forward = forward, .height = h - drop }, cam_focal);
}

fn lerp(a: camera.ScreenPt, b: camera.ScreenPt, t: f32) camera.ScreenPt {
    return .{ .x = a.x + (b.x - a.x) * t, .y = a.y + (b.y - a.y) * t };
}

// screen point of corner k at world height h: the cross-section shrinks linearly to
// the apex, so it's just base→apex lerp by h/HEIGHT.
fn ring(base_s: [4]camera.ScreenPt, apex_s: camera.ScreenPt, k: usize, h: f32) camera.ScreenPt {
    return lerp(base_s[k], apex_s, h / TOWER_HEIGHT);
}

// a thick screen-space line as a filled quad of width wpx.
fn bar(a: camera.ScreenPt, b: camera.ScreenPt, wpx: f32) void {
    const dx = b.x - a.x;
    const dy = b.y - a.y;
    var len = @sqrt(dx * dx + dy * dy);
    if (len < 1e-4) len = 1;
    const ox = -dy / len * wpx / 2.0;
    const oy = dx / len * wpx / 2.0;
    const pts = [_]camera.ScreenPt{
        .{ .x = a.x + ox, .y = a.y + oy },
        .{ .x = b.x + ox, .y = b.y + oy },
        .{ .x = b.x - ox, .y = b.y - oy },
        .{ .x = a.x - ox, .y = a.y - oy },
    };
    paint.pushPoly(TOWER_METAL, &pts);
}

/// drawFlat renders a tower given its four base corners + base centre already mapped
/// into the rider frame. Projects the corners + apex, then derives every ring and
/// brace by 2D interpolation. Sinks the tower by the local ground drop.
pub fn drawFlat(base: [4]geom.RiderPt, center: geom.RiderPt, cam_focal: f32) void {
    if (center.forward < camera.NEAR) return;
    const drop = groundDrop(center);
    if (drop >= TOWER_HEIGHT) return; // sunk below the horizon
    const clip_h = drop;

    var base_s: [4]camera.ScreenPt = undefined;
    for (base, 0..) |b, k| base_s[k] = proj(b.right, b.forward, 0, drop, cam_focal);
    const apex_s = proj(center.right, center.forward, TOWER_HEIGHT, drop, cam_focal);

    // rod width in px at this depth (project two ground points 1m apart).
    const px1 = camera.project(.{ .right = 1, .forward = center.forward, .height = 0 }, cam_focal).x;
    const px0 = camera.project(.{ .right = 0, .forward = center.forward, .height = 0 }, cam_focal).x;
    const wpx = ROD_W * (px1 - px0);

    // four legs (from the drop height up to the apex).
    var k: usize = 0;
    while (k < 4) : (k += 1) bar(ring(base_s, apex_s, k, clip_h), apex_s, wpx);

    // cross-beam rings at 20/40/60 (above the drop).
    var h: f32 = STAGE_HEIGHT;
    while (h < TOWER_HEIGHT) : (h += STAGE_HEIGHT) {
        if (h <= clip_h) continue;
        k = 0;
        while (k < 4) : (k += 1) bar(ring(base_s, apex_s, k, h), ring(base_s, apex_s, (k + 1) % 4, h), wpx);
    }

    // X-braces on the bottom stages, each diagonal clipped at the drop height.
    var stage: usize = 0;
    while (stage < BRACE_STAGES) : (stage += 1) {
        const lo = @as(f32, @floatFromInt(stage)) * STAGE_HEIGHT;
        const hi = lo + STAGE_HEIGHT;
        if (hi <= clip_h) continue;
        const f = (@max(lo, clip_h) - lo) / STAGE_HEIGHT;
        k = 0;
        while (k < 4) : (k += 1) {
            const j = (k + 1) % 4;
            bar(lerp(ring(base_s, apex_s, k, lo), ring(base_s, apex_s, j, hi), f), ring(base_s, apex_s, j, hi), wpx);
            bar(lerp(ring(base_s, apex_s, j, lo), ring(base_s, apex_s, k, hi), f), ring(base_s, apex_s, k, hi), wpx);
        }
    }

    drawBeacon(apex_s, center.forward, cam_focal);
}

// the apex beacon: a pink disc, size scaling with distance. Static for now (the
// step-driven blink + alpha glow ride the deferred animation clock).
fn drawBeacon(apex_s: camera.ScreenPt, forward: f32, cam_focal: f32) void {
    const px1 = camera.project(.{ .right = 1, .forward = forward, .height = 0 }, cam_focal).x;
    const px0 = camera.project(.{ .right = 0, .forward = forward, .height = 0 }, cam_focal).x;
    const r = BEACON_RADIUS * (px1 - px0);
    if (r < 0.5) return;
    var pts: [16]camera.ScreenPt = undefined;
    var i: usize = 0;
    while (i < 16) : (i += 1) {
        const a = @as(f32, @floatFromInt(i)) / 16.0 * 2.0 * std.math.pi;
        pts[i] = .{ .x = apex_s.x + r * @cos(a), .y = apex_s.y + r * @sin(a) };
    }
    paint.pushPoly(BEACON_COLOR, &pts);
}
